networking_generic_switch/devices/netmiko_devices/juniper.py

```diff
@@ -49,6 +54,44 @@ class Juniper(netmiko_devices.NetmikoSwitch):
         'vlan members {segmentation_id}',
     )
 
+    @tenacity.retry(reraise=True, stop=tenacity.stop_after_attempt(3),
+                    wait=tenacity.wait_fixed(14))
+    def send_config_set(self, net_connect, config_commands=None,
+                        exit_config_mode=True, delay_factor=1, max_loops=150,
+                        strip_prompt=False, strip_command=False,
+                        config_mode_cmd='configure private'):
+        """Temporarily overriding the send_config_set method from netmiko.
+
+        Temporarily overriding the send_config_set method from netmiko, until
+        upstream patch is accepted:
+
+        https://github.com/ktbyers/netmiko/pull/593
+
+        """
+
+        delay_factor = net_connect.select_delay_factor(delay_factor)
+        if config_commands is None:
+            return ''
+        elif isinstance(config_commands, string_types):
+            config_commands = (config_commands,)
+
+        if not hasattr(config_commands, '__iter__'):
+            raise ValueError("Invalid argument passed into send_config_set")
+
+        # Send config commands
+        output = net_connect.config_mode(config_mode_cmd)
+        for cmd in config_commands:
+            net_connect.write_channel(net_connect.normalize_cmd(cmd))
+            time.sleep(delay_factor * .5)
+
+        # Gather output
+        output += net_connect._read_channel_timing(
+            delay_factor=delay_factor, max_loops=max_loops)
+        if exit_config_mode:
+            output += net_connect.exit_config_mode()
+        output = net_connect._sanitize_output(output)
+        return output
```

Comment:
Would this work:

```
    @tenacity.retry(reraise=True, stop=tenacity.stop_after_attempt(3),
                    wait=tenacity.wait_fixed(14))
    def send_config_set(self, net_connect, config_commands=None):
        net_connect.config_mode(config_command='configure private')
        net_connect.send_config_set(config_commands)
```

Once set to private config mode, netmiko should not change it to a different config mode.

Reply:
There's a `self.config_mode()` call within `send_config_set` prior to any commands being sent, which presumable would reset the config_command back to it's defaults, which in the case of juniper, would be an empty string. Phil has suggested just monkey patching the config_mode, but it would still need to be called from a monkey-patched send_config_set so we may as well just override that, as I've done here.

Reply:
Right, but it's a noop if the switch is already in config mode:

```
     def config_mode(self, config_command='', pattern=''):
         """Enter into config_mode."""
         output = ''
         if not self.check_config_mode():
             self.write_channel(self.normalize_cmd(config_command))
             output = self.read_until_pattern(pattern=pattern)
             if not self.check_config_mode():
                 raise ValueError("Failed to enter configuration mode.")
         return output
```

Reply:
Actually, the monkey patching I'm suggesting would take place entirely in netmiko_devices.juniper.Juniper and amounts to overriding the superclass _get_connection().  Darren has my example code now to take a look at.